File: 4.8-bc/go/stage-1-flat/app/handler/subscription.go
package handler

import (
	"context"
	"encoding/json"
	"net/http"

	"github.com/go-chi/chi/v5"
	"github.com/google/uuid"

	"github.com/genkovich/claude-course-demos/4.8-bc/go/stage-1-flat/app/model"
)

type subscriptionServiceAPI interface {
	Subscribe(ctx context.Context, userID uuid.UUID, plan string) (*model.Subscription, error)
}

type subscriptionHandler struct {
	svc subscriptionServiceAPI
}

func MountSubscription(r chi.Router, svc subscriptionServiceAPI) {
	h := &subscriptionHandler{svc: svc}
	r.Post("/subscriptions", h.subscribe)
}

type subscribeReq struct {
	UserID string `json:"user_id"`
	Plan   string `json:"plan"`
}

func (h *subscriptionHandler) subscribe(w http.ResponseWriter, r *http.Request) {
	var req subscribeReq
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil || req.Plan == "" {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "invalid_request"})
		return
	}
	uid, err := uuid.Parse(req.UserID)
	if err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "invalid_user_id"})
		return
	}
	sub, err := h.svc.Subscribe(r.Context(), uid, req.Plan)
	if err != nil {
		writeJSON(w, http.StatusInternalServerError, map[string]string{"error": "subscribe_failed"})
		return
	}
	writeJSON(w, http.StatusCreated, map[string]any{
		"subscription_id": sub.ID.String(),
		"plan":            sub.Plan,
		"next_charge_at":  sub.NextChargeAt,
	})
}

File: 4.8-bc/ts/stage-1-flat/src/handler/subscription.ts
import type { FastifyInstance } from "fastify";

import type { SubscriptionService } from "../service/subscription.js";

interface SubscribeBody {
  user_id?: string;
  plan?: string;
}

const UUID_RE =
  /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;

export function mountSubscription(
  app: FastifyInstance,
  svc: SubscriptionService,
): void {
  app.post<{ Body: SubscribeBody }>("/subscriptions", async (req, reply) => {
    const body = req.body ?? {};
    if (!body.plan) {
      return reply.code(400).send({ error: "invalid_request" });
    }
    if (!body.user_id || !UUID_RE.test(body.user_id)) {
      return reply.code(400).send({ error: "invalid_user_id" });
    }
    try {
      const sub = await svc.subscribe(body.user_id, body.plan);
      return reply.code(201).send({
        subscription_id: sub.id,
        plan: sub.plan,
        next_charge_at: sub.nextChargeAt,
      });
    } catch (err) {
      req.log.error({ err }, "subscribe failed");
      return reply.code(500).send({ error: "subscribe_failed" });
    }
  });
}
